modules/images/webp-uploads/load.php

```diff
@@ -683,3 +683,190 @@ function webp_uploads_update_rest_attachment( WP_REST_Response $response, WP_Pos
+ *
+ * @param int   $attachment_id The ID of the attachment.
+ * @param array $data The current metadata to be stored in the attachment.
+ * @return array The updated metadata of the attachment.
+ */
+function webp_uploads_restore_image( $attachment_id, $data ) {
+	$backup_sizes = get_post_meta( $attachment_id, '_wp_attachment_backup_sizes', true );
+
+	if ( ! is_array( $backup_sizes ) ) {
+		return $data;
+	}
+
+	// If `IMAGE_EDIT_OVERWRITE` is defined and is truthy remove any edited images if present before replacing the metadata.
+	if ( defined( 'IMAGE_EDIT_OVERWRITE' ) && IMAGE_EDIT_OVERWRITE ) {
+		$file     = get_attached_file( $attachment_id );
+		$dirname  = pathinfo( $file, PATHINFO_DIRNAME );
+		$metadata = wp_get_attachment_metadata( $attachment_id );
+
+		$sources = isset( $metadata['sources'] ) && is_array( $metadata['sources'] ) ? $metadata['sources'] : array();
+
+		foreach ( $sources as $mime => $properties ) {
+			if ( empty( $properties['file'] ) ) {
+				continue;
+			}
+
+			// Delete only if it's an edited image.
+			if ( preg_match( '/-e\d{13}/', $properties['file'] ) ) {
+				$delete_file = path_join( $dirname, $properties['file'] );
+				wp_delete_file( $delete_file );
+			}
+		}
+
+		foreach ( $metadata['sizes'] as $size_name => $properties ) {
+			if ( ! isset( $properties['sources'] ) || ! is_array( $properties['sources'] ) ) {
+				continue;
+			}
+
+			foreach ( $properties['sources'] as $mime => $source_properties ) {
+				if ( empty( $source_properties['file'] ) ) {
+					continue;
+				}
+
+				// Delete only if it's an edited image.
+				if ( preg_match( '/-e\d{13}/', $source_properties['file'] ) ) {
+					$delete_file = path_join( $dirname, $source_properties['file'] );
+					wp_delete_file( $delete_file );
+				}
+			}
+		}
+	}
+
+	if ( isset( $backup_sizes['full-orig']['sources'] ) ) {
+		$data['sources'] = $backup_sizes['full-orig']['sources'];
+	}
+
+	return $data;
+}
+
+/**
+ * Hook fired right after a metadata has been created or updated, this function would look
+ * specifically only for the key: `_wp_attachment_backup_sizes` which is the one used to
+ * store all the backup sizes. This hook is in charge of cleaning up the additional meta
+ * keys stored in the metadata of the attachment and storing the `sources` property in the
+ * previous full size image due this is not a size we need to move the `sources` from the
+ * metadata back into the full size similar as how it's done on the rest of the sizes.
+ *
+ * @since n.e.x.t
+ *
+ * @param int    $meta_id The ID of the meta value stored in the DB.
+ * @param int    $attachment_id The ID of the post used for this meta, in our case the attachment ID.
+ * @param string $meta_name The name of the metadata to be updated.
+ * @param array  $backup_sizes An array with the metadata value in this case the backup sizes.
+ */
+function webp_updated_postmeta( $meta_id, $attachment_id, $meta_name, $backup_sizes ) {
```

Comment:
I don't think this function is necessary, and it makes the code overly complicated. We don't have to put the `sources` into `_wp_attachment_backup_sizes`. I see WP core doesn't do this by default, since it manually assembles the array for the `full-orig` size. However, a much simpler and less hacky implementation would be to rely on the separate `_wp_attachment_backup_sources` meta key I suggested above. What you have here may work, but it has some crazy filter logic that would be better to avoid, the approach of a separate meta key will be much simpler.

This way we don't need to bother with WP core's `_wp_attachment_backup_sizes` and can get rid of this entire function.

Reply:
Agree with this approach, the main idea behind this initial approach was to maximize what we already have and avoid introducing more meta data keys. Will update the code with the suggested approach instead.